src/ophyd_async/core/signal.py

```diff
@@ -66,13 +66,21 @@ def __init__(
         super().__init__(name)
 
     async def connect(
-        self, mock=False, timeout=DEFAULT_TIMEOUT, force_reconnect: bool = False
+        self,
+        mock=False,
+        timeout=DEFAULT_TIMEOUT,
+        force_reconnect: bool = False,
+        backend: Optional[SignalBackend[T]] = None,
     ):
+        if backend:
+            if self._initial_backend and backend != self._initial_backend:
+                raise ValueError(
+                    "Backend at connection different from initialised one."
+                )
+            self._initial_backend = backend
         if mock and not isinstance(self._backend, MockSignalBackend):
             # Using a soft backend, look to the initial value
-            self._backend = MockSignalBackend(
-                initial_backend=self._initial_backend,
-            )
+            self._backend = MockSignalBackend(initial_backend=self._initial_backend)
         self.log.debug(f"Connecting to {self.source}")
         await self._backend.connect(timeout=timeout)
```

Comment:
In the case where no backend was passed on init and it's not in mock, won't `self._backend` still be `None` here?

``` python
await self._backend.connect(timeout=timeout)
```

I think it would be better to keep `_initial_backend` as `None` and change `_backend` here.

``` python
            self._backend = backend
        if mock and not isinstance(self._backend, MockSignalBackend):
            # Perhaps a rename of
            # MockSignalBackend.initial_backend -> MockSignalBackend.real_backend?
            self._backend = MockSignalBackend(initial_backend=self._backend)
        self.log.debug(f"Connecting to {self.source}")
        await self._backend.connect(timeout=timeout)
```

Reply:
Yh good point, changed now